File: ggcommons/heartbeat/heartbeat.py
import logging
import threading
import time
from abc import ABC

from awsiot.greengrasscoreipc.model import QOS

from ggcommons.config.manager.configuration_change_listener import (
    ConfigurationChangeListener,
)
from ggcommons.heartbeat.heartbeat_monitor import HeartbeatMonitor
from ggcommons.messaging.message_builder import MessageBuilder
from ggcommons.messaging.messaging_client import MessagingClient
from ggcommons.metrics.metric_emitter import MetricEmitter
from ggcommons.metrics.metric import Metric
from ggcommons.metrics.measure import Measure
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ggcommons.config.manager.config_manager import ConfigManager

logger = logging.getLogger("Heartbeat")


class Heartbeat(ConfigurationChangeListener, ABC):
    __MESSAGE_NAME = "heartbeat"
    __MESSAGE_VERSION = "1.0.0"

    def __init__(self, config_service: "ConfigManager"):
        super().__init__()
        logger.info(f"Initializing Heartbeat system for component: {config_service.get_component_name()}")
        
        self._heartbeat_thread = None
        self._config_service = config_service
        self._config_service.add_config_change_listener(self)
        self._config = self._config_service.get_heartbeat_config()
        self.keep_running = True
        
        logger.info(f"Heartbeat configured - interval: {self._config.get_interval_secs()}s, targets: {len(self._config.get_targets())}")
        logger.debug(f"Heartbeat targets: {[target['type'] for target in self._config.get_targets()]}")
        
        self._define_metric(config_service)
        self.__run_heartbeat()
        
        logger.info("Heartbeat system initialization completed")

    def _define_metric(self, config_service: "ConfigManager"):
        logger.info("Defining heartbeat metrics")
        
        storage_resolution = (
            1
            if config_service.get_heartbeat_config().get_interval_secs() < 60
            else 60
        )
        
        logger.debug(f"Heartbeat metric storage resolution: {storage_resolution}s")
        
        from ggcommons.metrics.metric_builder import MetricBuilder
        
        measures = [
            ("disk_total", "Gigabytes"), ("disk_used", "Gigabytes"), ("disk_free", "Gigabytes"),
            ("cpu_usage", "Percent"), ("memory_usage", "Megabytes"), ("threads", "Count"),
            ("files", "Count"), ("fds", "Count")
        ]
        
        metric_builder = MetricBuilder.create("heartbeat") \
            .with_namespace(config_service.get_metric_config().get_namespace())
        
        for measure_name, unit in measures:
            metric_builder.add_measure(measure_name, unit, storage_resolution)
        
        metric = metric_builder.build()
        
        MetricEmitter.define_metric(metric)
        logger.info(f"Heartbeat metric defined with {len(measures)} measures")

    def __publish_heartbeat(self, heartbeat_monitor: HeartbeatMonitor):
        data = heartbeat_monitor.get_stats()
        logger.debug(f"Publishing heartbeat data with {len(data)} metrics")
        
        for target in self._config.get_targets():
            target_type = target["type"]
            logger.debug(f"Processing heartbeat target: {target_type}")
            
            if target_type == "messaging":
                message = MessageBuilder.create(
                    Heartbeat.__MESSAGE_NAME,
                    Heartbeat.__MESSAGE_VERSION
                ).with_payload(data).with_config(self._config_service).build()
                
                topic = self._config.DEFAULT_HEARTBEAT_MESSAGING_TOPIC
                destination = self._config.DEFAULT_HEARTBEAT_MESSAGING_DESTINATION
                
                if "config" in target:
                    if "topic" in target["config"]:
                        topic = self._config_service.resolve_template(
                            target["config"]["topic"]
                        )
                    if "destination" in target["config"]:
                        destination = target["config"]["destination"]
                
                logger.debug(f"Publishing heartbeat to {destination} destination on topic: {topic}")
                
                if destination.lower() == "ipc":
                    MessagingClient.publish(topic, message)
                else:
                    MessagingClient.publish_to_iot_core(
                        topic, message, QOS.AT_LEAST_ONCE
                    )
                    
            elif target_type == "metric":
                measure_values = {}
                for key, value in data.items():
                    if isinstance(value, dict):
                        for measure_name, measure_value in value.items():
                            measure_values[measure_name] = float(measure_value)
                
                logger.debug(f"Emitting heartbeat metrics with {len(measure_values)} measures")
                MetricEmitter.emit_metric_now("heartbeat", measure_values)

    def __heartbeat_loop(self):
        heartbeat_monitor = HeartbeatMonitor(self._config_service)
        interval = self._config_service.get_heartbeat_config().get_interval_secs()
        
        logger.info(f"Starting heartbeat loop with {interval}s interval")
        
        try:
            while self.keep_running:
                logger.debug("Heartbeat cycle starting")
                self.__publish_heartbeat(heartbeat_monitor)
                logger.debug(f"Heartbeat cycle completed, sleeping for {interval}s")
                time.sleep(interval)
            logger.info("Heartbeat loop stopped intentionally")
        except KeyboardInterrupt:
            logger.error("Heartbeat loop interrupted by user")
        except Exception as exc:
            logger.exception(f"Error in heartbeat loop: {exc}")

    def __run_heartbeat(self):
        try:
            thread_name = f"{self._config_service.get_component_name()}-heartbeat"
            logger.info(f"Starting heartbeat thread: {thread_name}")
            
            self._heartbeat_thread = threading.Thread(
                target=self.__heartbeat_loop,
                name=thread_name,
            )
            self._heartbeat_thread.daemon = True
            self._heartbeat_thread.start()
            
            logger.info(f"Heartbeat thread started successfully: {thread_name}")
            
        except Exception as exc:
            logger.error(f"Failed to start heartbeat thread: {exc}")
            raise

    def on_configuration_change(self, configuration) -> bool:
        logger.info("Heartbeat configuration changed, restarting heartbeat system")
        
        # Stop current heartbeat
        logger.debug("Stopping current heartbeat thread")
        self.keep_running = False
        self._heartbeat_thread.join()
        
        # Restart with new configuration
        logger.debug("Reloading heartbeat configuration")
        self.keep_running = True
        self._config = self._config_service.get_heartbeat_config()
        
        logger.info(f"Heartbeat reconfigured - new interval: {self._config.get_interval_secs()}s")
        
        self._define_metric(self._config_service)
        self.__run_heartbeat()
        
        logger.info("Heartbeat system restart completed")
        return True

    def stop(self):
        logger.info("Stopping heartbeat system")
        self.keep_running = False
        if self._heartbeat_thread:
            self._heartbeat_thread.join()
        logger.info("Heartbeat system stopped")

    def start(self):
        logger.info("Starting heartbeat system")
        self.keep_running = True
        self.__run_heartbeat()

File: libs/java/src/main/java/com/mbreissi/ggcommons/heartbeat/Heartbeat.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.ggcommons.heartbeat;

import com.mbreissi.ggcommons.config.ConfigManager;
import com.mbreissi.ggcommons.config.ConfigurationChangeListener;
import com.mbreissi.ggcommons.config.HeartbeatConfiguration;
import com.mbreissi.ggcommons.messaging.Message;
import com.mbreissi.ggcommons.messaging.MessageBuilder;
import com.mbreissi.ggcommons.messaging.MessagingClient;
import com.mbreissi.ggcommons.metrics.Metric;
import com.mbreissi.ggcommons.metrics.MetricBuilder;
import com.mbreissi.ggcommons.metrics.MetricEmitter;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import software.amazon.awssdk.aws.greengrass.model.QOS;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * Implements heartbeat functionality for Greengrass components to monitor their health status.
 * This class periodically publishes heartbeat messages and handles configuration changes.
 */
public class Heartbeat implements ConfigurationChangeListener
{
    protected static final Logger LOGGER = LogManager.getLogger(Heartbeat.class);

    private static final String MESSAGE_NAME = "heartbeat";
    private static final String MESSAGE_VERSION = "1.0.0";
    private final ConfigManager configurationService;
    private MessagingClient messagingService;
    private MetricEmitter metricService;
    private HeartbeatMonitor heartbeatMonitor;
    private final ScheduledExecutorService scheduler =
            Executors.newSingleThreadScheduledExecutor(runnable -> {
                Thread thread = new Thread(runnable, "Heartbeat-scheduler");
                thread.setDaemon(true);
                return thread;
            });
    private ScheduledFuture<?> heartbeatTask;
    private final Object timerLock = new Object();

    /**
     * Package-private constructor used by HeartbeatBuilder.
     * Use HeartbeatBuilder.create() instead of calling this directly.
     */
    Heartbeat(ConfigManager configurationService, MessagingClient messagingService, MetricEmitter metricService)
    {
        this.configurationService = configurationService;
        this.messagingService = messagingService;
        this.metricService = metricService;
        
        configurationService.addConfigChangeListener(this);
        initialize();
    }
    
    /**
     * Initializes the heartbeat after all dependencies are set.
     */
    private void initialize() {
        defineMetric();
        initHeartbeat();
    }

    /**
     * Initializes the heartbeat mechanism based on the current configuration.
     * Sets up the timer for periodic heartbeat publishing.
     */
    private void initHeartbeat()
    {
        synchronized (timerLock) {
            // Reschedule on the same executor: cancel the current task and submit a
            // new one at the configured interval. The task is a reusable Runnable, so
            // the executor is kept alive across reconfigures and only shut down in close().
            if (heartbeatTask != null) {
                heartbeatTask.cancel(false);
            }
            heartbeatMonitor = new HeartbeatMonitor(configurationService.getHeartbeatConfig());
            long periodMs = configurationService.getHeartbeatConfig().getIntervalSecs() * 1000L;
            heartbeatTask = scheduler.scheduleAtFixedRate(this::runHeartbeat, 0, periodMs, TimeUnit.MILLISECONDS);
            LOGGER.info("Heartbeat initialized at {} second interval", configurationService.getHeartbeatConfig().getIntervalSecs());
        }
    }

    /**
     * Defines the heartbeat metric in the metrics system.
     * This metric is used to track the component's health status.
     */
    private void defineMetric()
    {
        int storageResolution = configurationService.getHeartbeatConfig().getIntervalSecs() < 60 ? 1 : 60;
        Metric metric = MetricBuilder.create("heartbeat")
                .withNamespace(configurationService.getMetricConfig().getNamespace())
                .withConfig(configurationService)
                .addMeasure("disk_total", "Gigabytes", storageResolution)
                .addMeasure("disk_used", "Gigabytes", storageResolution)
                .addMeasure("disk_free", "Gigabytes", storageResolution)
                .addMeasure("cpu_usage", "Percent", storageResolution)
                .addMeasure("memory_usage", "Megabytes", storageResolution)
                .addMeasure("threads", "Count", storageResolution)
                .addMeasure("files", "Count", storageResolution)
                .addMeasure("fds", "Count", storageResolution)
                .build();
        metricService.defineMetric(metric);
    }

    /**
     * Publishes a heartbeat message to indicate the component is alive and functioning.
     * The message includes the current timestamp and component information.
     */
    private void publishHeartbeat()
    {
        JsonObject data = heartbeatMonitor.getStats();
        for (HeartbeatConfiguration.HeartbeatTarget target : configurationService.getHeartbeatConfig().getTargets())
        {
            switch (target.getType().toLowerCase())
            {
                case "metric":
                    var measureValues = new HashMap<String, Float>();
                    for (Map.Entry<String, JsonElement> entry : data.entrySet())
                    {
                        for (String measureName : entry.getValue().getAsJsonObject().keySet())
                        {
                            measureValues.put(measureName, entry.getValue().getAsJsonObject().get(measureName).getAsFloat());
                        }
                    }
                    metricService.emitMetricNow("heartbeat", measureValues);
                    break;

                case "messaging":
                    String topic = configurationService.resolveTemplate(HeartbeatConfiguration.DEFAULT_TOPIC);
                    String destination = HeartbeatConfiguration.DEFAULT_MESSAGING_DESTINATION;

                    if (target.getConfig().has("destination"))
                    {
                        destination = target.getConfig().get("destination").getAsString();
                    }
                    if (target.getConfig().has("topic"))
                    {
                        topic = configurationService.resolveTemplate(target.getConfig().get("topic").getAsString());
                    }

                    Message heartbeatMessage = MessageBuilder.create(MESSAGE_NAME, MESSAGE_VERSION)
                            .withPayload(heartbeatMonitor.getStats())
                            .withConfig(configurationService)
                            .build();
                    
                    // Canonical "ipc"/"iot_core"; the legacy "local"/"iotcore" spellings
                    // are accepted too (parity with the Python/Rust heartbeat targets).
                    if (destination.equalsIgnoreCase("ipc") || destination.equalsIgnoreCase("local"))
                    {
                        messagingService.publish(topic, heartbeatMessage);
                    }
                    else if (destination.equalsIgnoreCase("iot_core") || destination.equalsIgnoreCase("iotcore"))
                    {
                        messagingService.publishToIoTCore(topic, heartbeatMessage, QOS.AT_LEAST_ONCE);
                    }
                    else
                    {
                        LOGGER.warn("Unrecognized messaging destination: '{}'. Ignoring.", destination);
                    }
                    break;
            }
        }

    }

    /**
     * Stops the heartbeat, cancelling its periodic timer.
     */
    public void close()
    {
        synchronized (timerLock)
        {
            if (heartbeatTask != null)
            {
                heartbeatTask.cancel(false);
                heartbeatTask = null;
            }
            scheduler.shutdownNow();
        }
    }

    @Override
    /**
     * Handles configuration changes by reinitializing the heartbeat mechanism.
     * 
     * @return true if the configuration change was handled successfully
     */
    public boolean onConfigurationChanged()
    {
        LOGGER.info("Configuration changed, restarting heartbeat");
        initHeartbeat();
        return true;
    }

    /**
     * Inner class that implements the periodic heartbeat task.
     * Executes the heartbeat publishing operation at configured intervals.
     */
    /**
     * The periodic heartbeat task. Guarded so an exception in one run cannot
     * propagate to the scheduler and silently cancel future heartbeats.
     */
    private void runHeartbeat()
    {
        try
        {
            publishHeartbeat();
        }
        catch (Exception e)
        {
            LOGGER.error("Heartbeat task failed; will retry next interval", e);
        }
    }
}
